submitr/s3_upload.py

```diff
@@ -0,0 +1,322 @@
+from boto3 import client as BotoClient
+from collections import namedtuple
+import threading
+from time import time as current_timestamp
+from typing import Callable, Optional
+from dcicutils.file_utils import compute_file_md5
+from dcicutils.misc_utils import format_duration, format_size
+from dcicutils.progress_bar import ProgressBar
+from submitr.file_for_upload import FileForUpload
+from submitr.output import Question
+from submitr.rclone import RCloner, RCloneAmazon, cloud_path
+from submitr.s3_utils import get_s3_bucket_and_key_from_s3_uri, get_s3_key_metadata
+from submitr.utils import chars
+
+# Module to upload a given file, with the given AWS credentials to AWS S3.
+# Displays progress bar and other info; checks if file already exists; verifies
+# upload; catches interrupts; et cetera. Circa May 2024 added support for upload,
+# or transfer rather, from Google Cloud Storage (GCS) directly to S3 via rclone.
+
+# Notes on checksums:
```

Comment:
This type of documentation is great but because it's inline it's going to be inherently difficult to navigate/discover, it would be helpful to move this to the RST documentation at some point